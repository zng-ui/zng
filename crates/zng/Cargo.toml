[package]
name = "zng"
version = "0.16.6"
authors = ["The Zng Project Developers"]
edition = "2024"
license = "Apache-2.0 OR MIT"
readme = "README.md"
description = "Cross-platform GUI framework."
documentation = "https://zng-ui.github.io/doc/zng"
repository = "https://github.com/zng-ui/zng"
categories = ["gui"]
keywords = ["gui", "ui", "user-interface", "zng"]

[features]
default = ["dev"]

# Defaults recommended for the dev profile (debug builds).
#
# Recommended `Cargo.toml` setup:
#
# ```toml
# [dependencies]
# zng = { default-features = false }
#
# [features]
# default = ["dev"]
#
# # development features:
# dev = [
#   "zng/dev",
#   "zng/view_prebuilt",
# ]
#
# # release features:
# # use `cargo build --release --no-default-features --features release` to build
# release = [
#   "zng/window",
#   "zng/view",
#   "zng/ipc",
#   "zng/crash_handler",
#   "zng/...",
# ]
# ```
dev = [
    "ipc", 
    "view_software", 
    "crash_handler",
    "crash_handler_debug",
    "window", 
    "third_party", 
    "third_party_default",
    "checkerboard",
    "clipboard",
    "color_filter",
    "config",
    "settings_editor",
    "data_context",
    "data_view",
    "dialog",
    "drag_drop",
    "fs_watcher",
    "grid",
    "image",
    "markdown",
    "menu",
    "progress",
    "rule_line",
    "scroll",
    "slider",
    "toggle",
    "stack",
    "text_input",
    "tooltip",
    "undo",
    "wrap",
    "shortcut_text",
    "image_all",
    "button",
    "ansi_text",
    "dyn_app_extension",
    "inspector",
    "trace_widget",
    "trace_recorder",
    "var_type_names",
]

# **deprecated** use "dev"
debug_default = [
    "dev",
    "zng-app/debug_default",
    "zng-wgt-inspector?/debug_default",
    "zng-wgt-scroll/debug_default",
    "zng-wgt-window/debug_default",
]

# Include the default view-process implementation.
#
# Only enables in `not(target_arch = "wasm32")` builds.
view = ["dep:zng-view"]
# Include the default view-process implementation as an embedded precompiled binary.
#
# Only enables in `not(any(target_arch = "wasm32", target_os = "android", target_os = "ios"))` builds.
view_prebuilt = ["dep:zng-view-prebuilt", "ipc"]

# Enables HTTP tasks and web features of widgets and services.
http = [
    "zng-task/http",
    "zng-wgt/http",
    "zng-ext-image?/http",
    "zng-ext-window?/http",
]

# Enable SVG image rendering, SVG emoji support.
svg = ["dep:zng-ext-svg", "zng-ext-font/svg"]

# **deprecated** no longer needed
dyn_node = []

# Instrument each property and widget instance with "Inspector" nodes and
# extend windows to be inspected on Ctrl+Shift+I.
inspector = ["dep:zng-wgt-inspector", "zng-app/inspector", "zng-wgt-inspector/live"]

# Enable hot reload builds.
#
# Note that you must configure the target library to hot reload, see `zng::hot_reload` for details.
#
# Only enables in `not(any(target_arch = "wasm32", target_os = "android", target_os = "ios"))` builds.
hot_reload = ["dep:zng-ext-hot-reload"]

# Use dynamic dispatch at the app-extension level.
#
# This speeds-up compilation time at the cost of runtime.
dyn_app_extension = ["zng-app/dyn_app_extension"]

# **deprecated** no longer needed
dyn_closure = []

# Test utilities.
test_util = [
    "zng-app/test_util",
    "zng-ext-window?/test_util",
    "zng-task/test_util",
]

# Allows multiple app instances per-process.
#
# This feature allows multiple apps, one app per thread at a time. The `LocalContext` tracks
# what app is currently running in each thread and `app_local!` statics switch to the value of each app
# depending on the current thread.
#
# Not enabled by default, but enabled by `feature="test_util"`.
multi_app = ["zng-app/multi_app"]

# Enables single app-process instance mode.
#
# Builds with this feature only allow one app-process, subsequent attempts to spawn the app redirect to
# the running app-process.
#
# Only enables in `not(any(target_arch = "wasm32", target_os = "android", target_os = "ios"))` builds.
single_instance = ["dep:zng-ext-single-instance"]

# Allow app-process crash handler.
#
# Builds with this feature spawn a crash monitor-process for each app-process.
#
# Only enables in `not(any(target_arch = "wasm32", target_os = "android"))` builds.
crash_handler = ["zng-app/crash_handler"]

# Enable debug crash handler view.
crash_handler_debug = ["zng-wgt-inspector/crash_handler", "window"]

# Instrument every widget outer-most node to trace UI methods.
trace_widget = ["zng-app/trace_widget", "zng-wgt-style/trace_widget"]

# Enable trace recording.
#
# Note that this does not auto start recording, to do that run with the `ZNG_RECORD_TRACE` env var set.
trace_recorder = ["zng-app/trace_recorder"]

# Instrument every property and intrinsic node to trace UI methods.
#
# Note that this can cause very large trace files and bad performance.
trace_wgt_item = ["zng-app/trace_wgt_item"]

# Spawns a thread on app creation that checks and prints `parking_lot` deadlocks.
deadlock_detection = ["zng-app/deadlock_detection"]

# Embed hyphenation dictionaries for all supported languages.
#
# If enabled some 2.8MB of data is embedded, you can provide an alternative dictionary source using the 
# `HYPHENATION::init_data_source` method.
hyphenation_embed_all = ["zng-ext-font/hyphenation_embed_all"]

# Include all Material Icons icon sets in the default app.
material_icons = [
    "material_icons_outlined",
    "material_icons_filled",
    "material_icons_rounded",
    "material_icons_sharp",
]
# Material Icons Outlined icon set.
#
# If enabled some icons of this set are used for some of the commands.
material_icons_outlined = [
    "dep:zng-wgt-material-icons",
    "zng-wgt-material-icons/embedded",
    "zng-wgt-material-icons/outlined",
]

# Material Icons Filled icon set.
material_icons_filled = [
    "dep:zng-wgt-material-icons",
    "zng-wgt-material-icons/embedded",
    "zng-wgt-material-icons/filled",
]

# Material Icons Rounded icon set.
material_icons_rounded = [
    "dep:zng-wgt-material-icons",
    "zng-wgt-material-icons/embedded",
    "zng-wgt-material-icons/rounded",
]

# Material Icons Sharp icon set.
material_icons_sharp = [
    "dep:zng-wgt-material-icons",
    "zng-wgt-material-icons/embedded",
    "zng-wgt-material-icons/sharp",
]

# Compile with var value type names for better error messages and diagnostics.
var_type_names = ["zng-var/type_names", "zng-wgt-data?/var_type_names"]

# Enable JSON config format and file watcher helper.
config_json = ["zng-ext-config/json"]
# Enable TOML config format and file watcher helper.
config_toml = ["zng-ext-config/toml"]
# Enable RON config format and file watcher helper.
config_ron = ["zng-ext-config/ron"]
# Enable YAML config format and file watcher helper.
config_yaml = ["zng-ext-config/yaml"]
# Enable all config formats an file watcher helpers.
config_all = ["config_json", "config_toml", "config_ron", "config_yaml"]


# Enables software renderer fallback in the default view-process.
#
# If enabled and a native OpenGL 3.2 driver is not available the `swgl` software renderer is used.
view_software = ["zng-view?/software"]
# Collects and bundles third-party licenses used by the `zng-view` crate.
#
# Needs `cargo-about` and Internet connection during build.
#
# Not enabled by default. Note that `"view_prebuilt"` always bundles licenses.
view_bundle_licenses = ["zng-view?/bundle_licenses"]

# Enables IPC tasks, pre-build views and connecting to views running in another process.
#
# Only enables in `not(any(target_os = "android", target_arch = "wasm32", target_os = "ios"))` builds.
ipc = ["zng-app/ipc", "zng-view-api/ipc", "zng-view?/ipc", "zng-task/ipc"]

# Check if `zng::env::res` path is available in `init_built_res` first.
#
# Enabled by default in debug builds, ignored in Android and Wasm.
built_res = ["zng-env/built_res"]

# Standard Android backend that requires a build system that can compile Java or Kotlin and fetch Android dependencies.
#
# See `https://docs.rs/winit/latest/winit/platform/android/` for more details.
android_game_activity = ["zng-view?/android_game_activity"]

# Basic Android backend that does not require Java.
#
# See `https://docs.rs/winit/latest/winit/platform/android/` for more details. 
android_native_activity = ["zng-view?/android_native_activity"]

# Enable window, monitor services, widgets and properties.
window = ["dep:zng-ext-window", "dep:zng-wgt-window"]

# Enable third-party license service and types.
third_party = []

# Enable default third-party licenses default view.
third_party_default = ["window", "scroll", "text_input", "data_view", "markdown", "stack", "toggle"]

# Enable ANSI text widget.
#
# Not enabled by default.
ansi_text = ["dep:zng-wgt-ansi-text"]

# Enable checkerboard widget.
checkerboard = ["dep:zng-wgt-checkerboard"]

# Enable clipboard service.
clipboard = ["dep:zng-ext-clipboard"]

# Enable color filter properties.
color_filter = ["dep:zng-wgt-filter"]

# Enable file system watcher service.
fs_watcher = ["dep:zng-ext-fs-watcher"]

# Enable the configuration service.
config = ["dep:zng-ext-config", "fs_watcher"]

# Enable settings widgets.
settings_editor = ["dep:zng-wgt-settings"]

# Enable data context service and properties.
data_context = ["dep:zng-wgt-data"]

# Enable data view widget.
data_view = ["dep:zng-wgt-data-view"]

# Enable modal dialog overlay widget and service.
dialog = ["dep:zng-wgt-dialog"]

# Enable drag&drop.
drag_drop = ["zng-ext-input/drag_drop", "zng-wgt-input/drag_drop"]

# Enable grid widget.
grid = ["dep:zng-wgt-grid"]

# Enable image service and widget.
image = ["dep:zng-ext-image", "dep:zng-wgt-image"]

# Enable markdown widget.
markdown = ["dep:zng-wgt-markdown"]

# Enable menu widgets.
menu = ["dep:zng-wgt-menu"]

# Enable progress indicator widgets.
progress = ["dep:zng-wgt-progress"]

# Enable rule line widgets.
rule_line = ["dep:zng-wgt-rule-line"]

# Enable scroll widget.
scroll = ["dep:zng-wgt-scroll"]

# Enable button widget.
button = ["dep:zng-wgt-button"]

# Enable toggle widgets.
toggle = ["dep:zng-wgt-toggle"]

# Enable slider widget.
slider = ["dep:zng-wgt-slider"]

# Enable stack widget.
stack = ["dep:zng-wgt-stack"]

# Enable text input widgets.
text_input = ["dep:zng-wgt-text-input"]

# Enable tooltip widget.
tooltip = ["dep:zng-wgt-tooltip", "zng-wgt-button?/tooltip"]

# Enable undo/redo service.
undo = ["dep:zng-ext-undo", "dep:zng-wgt-undo-history", "dep:zng-wgt-undo"]

# Enable wrap widget.
wrap = ["dep:zng-wgt-wrap"]

# Enable the shortcut text widget.
shortcut_text = ["dep:zng-wgt-shortcut"]

# Enable BMP image decoder and encoder with "view" feature.
image_bmp = ["zng-view?/image_bmp"]
# Enable DDS image decoder with "view" feature.
image_dds = ["zng-view?/image_dds"]
# Enable EXR image decoder and encoder with "view" feature.
image_exr = ["zng-view?/image_exr"]
# Enable Farbfeld image decoder and encoder with "view" feature.
image_ff = ["zng-view?/image_ff"]
# Enable GIF image decoder and encoder with "view" feature.
image_gif = ["zng-view?/image_gif"]
# Enable Radiance HDR image decoder and encoder with "view" feature.
image_hdr = ["zng-view?/image_hdr"]
# Enable ICO image decoder and encoder with "view" feature.
image_ico = ["zng-view?/image_ico"]
# Enable JPEG image decoder and encoder with "view" feature.
image_jpeg = ["zng-view?/image_jpeg"]
# Enable PNG image decoder and encoder with "view" feature.
image_png = ["zng-view?/image_png"]
# Enable PNM image decoder and encoder with "view" feature.
image_pnm = ["zng-view?/image_pnm"]
# Enable QOI image decoder and encoder with "view" feature.
image_qoi = ["zng-view?/image_qoi"]
# Enable TGA image decoder and encoder with "view" feature.
image_tga = ["zng-view?/image_tga"]
# Enable TIFF image decoder and encoder with "view" feature.
image_tiff = ["zng-view?/image_tiff"]
# Enable WEBP image decoder with "view" feature.
image_webp = ["zng-view?/image_webp"]

# Enable all encoders and decoders.
image_all = ["zng-view?/image_all"]

[dependencies]
# foundation
zng-clone-move = { path = "../zng-clone-move", version = "0.3.1", default-features = false }
zng-txt = { path = "../zng-txt", version = "0.4.1", default-features = false }
zng-state-map = { path = "../zng-state-map", version = "0.6.4", default-features = false }
zng-app-context = { path = "../zng-app-context", version = "0.8.5", default-features = false }
zng-layout = { path = "../zng-layout", version = "0.5.5", default-features = false }
zng-var = { path = "../zng-var", version = "0.8.5", default-features = false }
zng-task = { path = "../zng-task", version = "0.7.6", default-features = false }
zng-color = { path = "../zng-color", version = "0.7.5", default-features = false }
zng-env = { path = "../zng-env", version = "0.6.4", default-features = false }
zng-unique-id = { path = "../zng-unique-id", version = "0.7.4", default-features = false }

# app
zng-app = { path = "../zng-app", version = "0.17.6", default-features = false }
zng-ext-fs-watcher = { path = "../zng-ext-fs-watcher", version = "0.6.6", optional = true, default-features = false }
zng-ext-config = { path = "../zng-ext-config", version = "0.7.6", default-features = false, optional = true }
zng-ext-font = { path = "../zng-ext-font", version = "0.8.6", default-features = false }
zng-ext-l10n = { path = "../zng-ext-l10n", version = "0.9.6", features = ["tar"], default-features = false }
zng-ext-image = { path = "../zng-ext-image", version = "0.6.6", optional = true, default-features = false }
zng-ext-svg = { path = "../zng-ext-svg", version = "0.5.6", optional = true, default-features = false }
zng-ext-clipboard = { path = "../zng-ext-clipboard", version = "0.6.6", optional = true, default-features = false }
zng-ext-window = { path = "../zng-ext-window", version = "0.7.6", optional = true, default-features = false }
zng-ext-input = { path = "../zng-ext-input", version = "0.9.6", default-features = false }
zng-ext-undo = { path = "../zng-ext-undo", version = "0.6.6", optional = true, default-features = false }
zng-ext-hot-reload = { path = "../zng-ext-hot-reload", version = "0.5.6", optional = true, default-features = false }

# widgets
zng-wgt = { path = "../zng-wgt", version = "0.9.6", default-features = false }
zng-wgt-access = { path = "../zng-wgt-access", version = "0.6.6", default-features = false }
zng-wgt-transform = { path = "../zng-wgt-transform", version = "0.6.6", default-features = false }
zng-wgt-input = { path = "../zng-wgt-input", version = "0.6.6", default-features = false }
zng-wgt-data = { path = "../zng-wgt-data", version = "0.6.6", optional = true, default-features = false }
zng-wgt-filter = { path = "../zng-wgt-filter", version = "0.6.6", optional = true, default-features = false }
zng-wgt-size-offset = { path = "../zng-wgt-size-offset", version = "0.6.6", default-features = false }
zng-wgt-container = { path = "../zng-wgt-container", version = "0.8.6", default-features = false }
zng-wgt-undo = { path = "../zng-wgt-undo", version = "0.6.6", optional = true, default-features = false }
zng-wgt-data-view = { path = "../zng-wgt-data-view", version = "0.7.6", optional = true, default-features = false }
zng-wgt-fill = { path = "../zng-wgt-fill", version = "0.6.6", default-features = false }
zng-wgt-style = { path = "../zng-wgt-style", version = "0.7.6", default-features = false }
zng-wgt-checkerboard = { path = "../zng-wgt-checkerboard", version = "0.7.6", optional = true, default-features = false }
zng-wgt-window = { path = "../zng-wgt-window", version = "0.10.6", optional = true, default-features = false }
zng-wgt-layer = { path = "../zng-wgt-layer", version = "0.7.6", default-features = false }
zng-wgt-undo-history = { path = "../zng-wgt-undo-history", version = "0.7.6", optional = true, default-features = false }
zng-wgt-image = { path = "../zng-wgt-image", version = "0.7.6", optional = true, default-features = false }
zng-wgt-text = { path = "../zng-wgt-text", version = "0.8.6", default-features = false }
zng-wgt-text-input = { path = "../zng-wgt-text-input", version = "0.8.6", optional = true, default-features = false }
zng-wgt-button = { path = "../zng-wgt-button", version = "0.8.6", default-features = false, optional = true }
zng-wgt-stack = { path = "../zng-wgt-stack", version = "0.7.6", optional = true, default-features = false }
zng-wgt-panel = { path = "../zng-wgt-panel", version = "0.7.6", default-features = false }
zng-wgt-grid = { path = "../zng-wgt-grid", version = "0.7.6", optional = true, default-features = false }
zng-wgt-wrap = { path = "../zng-wgt-wrap", version = "0.7.6", optional = true, default-features = false }
zng-wgt-shortcut = { path = "../zng-wgt-shortcut", version = "0.1.1", optional = true, default-features = false }
zng-wgt-rule-line = { path = "../zng-wgt-rule-line", version = "0.7.6", optional = true, default-features = false }
zng-wgt-toggle = { path = "../zng-wgt-toggle", version = "0.7.6", optional = true, default-features = false }
zng-wgt-menu = { path = "../zng-wgt-menu", version = "0.7.6", optional = true, default-features = false }
zng-wgt-scroll = { path = "../zng-wgt-scroll", version = "0.8.7", optional = true, default-features = false }
zng-wgt-ansi-text = { path = "../zng-wgt-ansi-text", version = "0.7.6", optional = true, default-features = false }
zng-wgt-tooltip = { path = "../zng-wgt-tooltip", version = "0.8.6", optional = true, default-features = false }
zng-wgt-markdown = { path = "../zng-wgt-markdown", version = "0.7.6", optional = true, default-features = false }
zng-wgt-inspector = { path = "../zng-wgt-inspector", version = "0.6.6", optional = true, default-features = false }
zng-wgt-settings = { path = "../zng-wgt-settings", version = "0.5.6", optional = true, default-features = false }
zng-wgt-dialog = { path = "../zng-wgt-dialog", version = "0.5.6", optional = true, default-features = false }
zng-wgt-progress = { path = "../zng-wgt-progress", version = "0.5.6", optional = true, default-features = false }
zng-wgt-slider = { path = "../zng-wgt-slider", version = "0.5.6", optional = true, default-features = false }

zng-wgt-material-icons = { path = "../zng-wgt-material-icons", version = "0.7.6", default-features = false, optional = true }
zng-ext-single-instance = { path = "../zng-ext-single-instance", version = "0.7.6", optional = true, default-features = false }

tracing = { version = "0.1", default-features = false, features = ["std"] }

# view
zng-view-api = { path = "../zng-view-api", version = "0.14.5", default-features = false, features = ["var"] }

[target.'cfg(not(target_arch = "wasm32"))'.dependencies]
zng-view = { path = "../zng-view", version = "0.11.5", default-features = false, optional = true }

[target.'cfg(not(any(target_arch = "wasm32", target_os = "android", target_os = "ios")))'.dependencies]
zng-view-prebuilt = { path = "../zng-view-prebuilt", version = "0.16.6", optional = true, default-features = false, features = ["embedded"] }

[build-dependencies]
cfg_aliases = { version = "0.2", default-features = false }

[package.metadata.docs.rs]
all-features = true

[package.metadata.zng-do]
# this file is generated on build, docs.rs does not give write access to target/doc.
rustdoc-args = [
    "--html-in-header",
    "target/doc/zng-material-icons-extensions.html",
]
