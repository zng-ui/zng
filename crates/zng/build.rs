#![recursion_limit = "512"]

fn main() {
    if cfg!(feature = "dyn_node") {
        println!(r#"cargo:warning=feature "dyn_node" is deprecated, no longer needed"#);
    }
    if cfg!(feature = "dyn_closure") {
        println!(r#"cargo:warning=feature "dyn_closure" is deprecated, no longer needed"#);
    }
    if cfg!(feature = "debug_default") {
        println!(r#"cargo:warning=feature "debug_default" is deprecated, enable needed features directly"#);
    }

    cfg_aliases::cfg_aliases! {
        wasm: { target_arch = "wasm32" },
        android: { target_os = "android" },
        ipc: { all(feature = "ipc", not(any(android, wasm))) },
        view: { all(feature = "view", not(wasm)) },
        view_prebuilt: { all(feature = "view_prebuilt", not(any(android, wasm))) },
        hot_reload: { all(feature = "hot_reload", not(any(android, wasm))) },
        single_instance: { all(feature = "single_instance", not(any(android, wasm))) },
        crash_handler: { all(feature = "crash_handler", not(any(android, wasm))) },
        trace_recorder: { all(feature = "trace_recorder", not(any(android, wasm))) },
    }
}
