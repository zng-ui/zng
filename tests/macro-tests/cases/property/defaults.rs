use zng::prelude_wgt::{property, IntoUiNode, IntoVar, UiNode};

#[property(CONTEXT, default)]
pub fn missing_default_parentheses(child: impl IntoUiNode, a: impl IntoVar<bool>, b: impl IntoVar<u32>) -> UiNode {
    let _ = (a, b);
    child.into_node()
}

#[property(CONTEXT, default())]
pub fn missing_default_values(child: impl IntoUiNode, a: impl IntoVar<bool>, b: impl IntoVar<u32>) -> UiNode {
    let _ = (a, b);
    child.into_node()
}

#[property(CONTEXT, default(true))]
pub fn incorrect_default_args_count_u_1(child: impl IntoUiNode, a: impl IntoVar<bool>, b: impl IntoVar<u32>) -> UiNode {
    let _ = (a, b);
    child.into_node()
}

#[property(CONTEXT, default(a: true))]
pub fn incorrect_default_args_count_n_1(child: impl IntoUiNode, a: impl IntoVar<bool>, b: impl IntoVar<u32>) -> UiNode {
    let _ = (a, b);
    child.into_node()
}

#[property(CONTEXT, default(true, 2555, "ABC"))]
pub fn incorrect_default_args_count_u_2(child: impl IntoUiNode, a: impl IntoVar<bool>, b: impl IntoVar<u32>) -> UiNode {
    let _ = (a, b);
    child.into_node()
}

#[property(CONTEXT, default(a: true, b: 2555, c: "ABC"))]
pub fn incorrect_default_args_count_n_2(child: impl IntoUiNode, a: impl IntoVar<bool>, b: impl IntoVar<u32>) -> UiNode {
    let _ = (a, b);
    child.into_node()
}

#[property(CONTEXT, default(2555, true))]
pub fn invalid_default_args_types_u_2(child: impl IntoUiNode, a: impl IntoVar<bool>, b: impl IntoVar<u32>) -> UiNode {
    let _ = (a, b);
    child.into_node()
}

#[property(CONTEXT, default(a: 2555, b: true))]
pub fn invalid_default_args_types_n_2(child: impl IntoUiNode, a: impl IntoVar<bool>, b: impl IntoVar<u32>) -> UiNode {
    let _ = (a, b);
    child.into_node()
}

fn main() {}
