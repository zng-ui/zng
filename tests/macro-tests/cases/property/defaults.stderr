error: unexpected end of input, expected parentheses
 --> cases/property/defaults.rs:3:1
  |
3 | #[property(CONTEXT, default)]
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  |
  = note: this error originates in the attribute macro `property` (in Nightly builds, run with -Z macro-backtrace for more info)

error: expected `,`
  --> cases/property/defaults.rs:21:30
   |
21 | #[property(CONTEXT, default(a: true))]
   |                              ^

error: expected `,`
  --> cases/property/defaults.rs:33:30
   |
33 | #[property(CONTEXT, default(a: true, b: 2555, c: "ABC"))]
   |                              ^

error: expected `,`
  --> cases/property/defaults.rs:45:30
   |
45 | #[property(CONTEXT, default(a: 2555, b: true))]
   |                              ^

error[E0061]: this method takes 2 arguments but 0 arguments were supplied
  --> cases/property/defaults.rs:9:21
   |
9  | #[property(CONTEXT, default())]
   |                     ^^^^^^^ two arguments are missing
   |
note: method defined here
  --> cases/property/defaults.rs:9:1
   |
9  | #[property(CONTEXT, default())]
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
10 | pub fn missing_default_values(child: impl UiNode, a: impl IntoVar<bool>, b: impl IntoVar<u32>) -> impl UiNode {
   |                                                   ---------------------  --------------------
   = note: this error originates in the attribute macro `property` (in Nightly builds, run with -Z macro-backtrace for more info)
help: provide the arguments
   |
9  | #[property(CONTEXT, default(/* a */, /* b */)())]
   |                            ++++++++++++++++++

error[E0061]: this method takes 2 arguments but 1 argument was supplied
  --> cases/property/defaults.rs:15:21
   |
15 | #[property(CONTEXT, default(true))]
   |                     ^^^^^^^ argument #2 is missing
   |
note: method defined here
  --> cases/property/defaults.rs:15:1
   |
15 | #[property(CONTEXT, default(true))]
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
16 | pub fn incorrect_default_args_count_u_1(child: impl UiNode, a: impl IntoVar<bool>, b: impl IntoVar<u32>) -> impl UiNode {
   |                                                                                    --------------------
   = note: this error originates in the attribute macro `property` (in Nightly builds, run with -Z macro-backtrace for more info)
help: provide the argument
   |
15 | #[property(CONTEXT, default(true, /* b */)(true))]
   |                            +++++++++++++++

error[E0061]: this method takes 2 arguments but 3 arguments were supplied
  --> cases/property/defaults.rs:27:21
   |
27 | #[property(CONTEXT, default(true, 2555, "ABC"))]
   |                     ^^^^^^^             ----- unexpected argument #3 of type `&'static str`
   |
note: method defined here
  --> cases/property/defaults.rs:27:1
   |
27 | #[property(CONTEXT, default(true, 2555, "ABC"))]
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   = note: this error originates in the attribute macro `property` (in Nightly builds, run with -Z macro-backtrace for more info)
help: remove the extra argument
   |
27 - #[property(CONTEXT, default(true, 2555, "ABC"))]
27 + #[property(CONTEXT, default(true, 2555))]
   |

error[E0277]: the trait bound `{integer}: IntoVar<bool>` is not satisfied
  --> cases/property/defaults.rs:39:29
   |
39 | #[property(CONTEXT, default(2555, true))]
   |                     ------- ^^^^ the trait `IntoVar<bool>` is not implemented for `{integer}`
   |                     |
   |                     required by a bound introduced by this call
   |
   = note: `IntoVar<T>` is implemented for all `T: VarValue`
   = note: `IntoVar<T>` is implemented for all `V: Var<T>`
   = help: the following other types implement trait `IntoVar<T>`:
            <implementers-list>
note: required by a bound in `invalid_default_args_types_u_2_::args`
  --> cases/property/defaults.rs:40:67
   |
39 | #[property(CONTEXT, default(2555, true))]
   | ----------------------------------------- required by a bound in this associated function
40 | pub fn invalid_default_args_types_u_2(child: impl UiNode, a: impl IntoVar<bool>, b: impl IntoVar<u32>) -> impl UiNode {
   |                                                                   ^^^^^^^^^^^^^ required by this bound in `invalid_default_args_types_u_2_::args`

error[E0277]: the trait bound `bool: IntoVar<u32>` is not satisfied
  --> cases/property/defaults.rs:39:35
   |
39 | #[property(CONTEXT, default(2555, true))]
   |                     -------       ^^^^ the trait `IntoVar<u32>` is not implemented for `bool`
   |                     |
   |                     required by a bound introduced by this call
   |
   = note: `IntoVar<T>` is implemented for all `T: VarValue`
   = note: `IntoVar<T>` is implemented for all `V: Var<T>`
   = help: the following other types implement trait `IntoVar<T>`:
            <implementers-list>
note: required by a bound in `invalid_default_args_types_u_2_::args`
  --> cases/property/defaults.rs:40:90
   |
39 | #[property(CONTEXT, default(2555, true))]
   | ----------------------------------------- required by a bound in this associated function
40 | pub fn invalid_default_args_types_u_2(child: impl UiNode, a: impl IntoVar<bool>, b: impl IntoVar<u32>) -> impl UiNode {
   |                                                                                          ^^^^^^^^^^^^ required by this bound in `invalid_default_args_types_u_2_::args`
